### PR TITLE
Kotlin migration

### PR DESCRIPTION
**Pull request by rekire**

### CODE REVIEW

**Comment by rekire** [comment]
Found no issues with testing feel free to check it yourself @brianegan

File: android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt
Comment (rekire) [reply android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt:11]: I was thinking to convert it, but it is a bit controversial what is better, so I kept the original way.

File: android/src/main/kotlin/vn/hunghd/flutterdownloader/FlutterDownloaderInitializer.kt
Comment (rekire) [reply android/src/main/kotlin/vn/hunghd/flutterdownloader/FlutterDownloaderInitializer.kt]: done

**Comment by rekire** [comment]
Please do not merge it yet, I noticed that I commited my changes with the wrong mail address. I need to fix that and force push it before you can merge it.

**Comment by bartekpacia** [comment]
@rekire Sure, no problem.

File: android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt
Comment (bartekpacia) [reply android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt:11]: >  it is a bit controversial what is better

Is it? Any source for this?

To me it seems like a perfect case for an enum.

Comment (rekire) [reply android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt:11]: Here is a random article about that topic: https://betterprogramming.pub/android-how-enums-can-impact-the-performance-f787ef5903dd

I have there no strong feelings I can convert this to an enum too. That will just cause some more changes.

By the way I fixed my mail address, but messed up the commit dates 😐, however that's not important for me

Comment (bartekpacia) [reply android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt:11]: Ah, you mean this problem with enums. That article is 3 years ago, but even then it was incorrect -  R8 has been optimizing enums for a good few years now, there's no performance penalty at all.

Comment (bartekpacia) [reply android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt:11]: So yeah, let's go with an enum.

Comment (rekire) [reply android/src/main/kotlin/vn/hunghd/flutterdownloader/DownloadStatus.kt:11]: I was not aware of this. I'll change that tomorrow

**Comment by rekire** [comment]
Can you verify that everything still works fine?

File: android/build.gradle
Comment (bartekpacia) [inline android/build.gradle:22]: what is this for?

File: android/build.gradle
Comment (bartekpacia) [inline android/build.gradle:54]: is this needed?

File: example/android/app/build.gradle
Comment (bartekpacia) [inline example/android/app/build.gradle:36]: Let's bump this in a separate PR.

File: example/android/app/build.gradle
Comment (bartekpacia) [inline example/android/app/build.gradle:48]: remove this change

File: android/src/main/AndroidManifest.xml
Comment (bartekpacia) [inline android/src/main/AndroidManifest.xml:10]: Adding this permission to the Manifest is not enough. We also need to show the permission request dialog to the user and handle the user's response.
It's already tracked in #710, so let's remove this line.

File: android/build.gradle
Comment (bartekpacia) [inline android/build.gradle:27]: Is this necessary for us to use Kotlin? If no, let's remove this.

File: android/src/main/kotlin/vn/hunghd/flutterdownloader/FlutterDownloaderInitializer.kt
Comment (bartekpacia) [inline android/src/main/kotlin/vn/hunghd/flutterdownloader/FlutterDownloaderInitializer.kt:46]: can we make these methods one-liners by using `=`?

**Review by bartekpacia — CHANGES_REQUESTED** [review]
Great work, I have just a few small remarks.

**Comment by bartekpacia** [comment]
Also, after cloning your branch and trying to use in my private app, I get the following crash:

<details>
    <summary>Build crash logs</summary>

    Launching lib/main.dart on sdk gphone64 arm64 in debug mode...
    lib/main.dart:1

    FAILURE: Build failed with an exception.

    * What went wrong:
    Execution failed for task ':flutter_downloader:generateDebugRFile'.
    > Could not resolve all files for configuration ':flutter_downloader:debugCompileClasspath'.
    > Could not resolve io.flutter:flutter_embedding_debug:1.0.0-ad3d868e0d174f5a8447b011bd714d18a5f2abec.
        Required by:
            project :flutter_downloader
        > Cannot find a version of 'io.flutter:flutter_embedding_debug' that satisfies the version constraints:
            Dependency path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:1.0.0-ad3d868e0d174f5a8447b011bd714d18a5f2abec'
            Dependency path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3'
            Constraint path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:{strictly 1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3}' because of the following reason: debugRuntimeClasspath uses version 1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3

    > Could not resolve io.flutter:flutter_embedding_debug:1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3.
        Required by:
            project :flutter_downloader
        > Cannot find a version of 'io.flutter:flutter_embedding_debug' that satisfies the version constraints:
            Dependency path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:1.0.0-ad3d868e0d174f5a8447b011bd714d18a5f2abec'
            Dependency path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3'
            Constraint path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:{strictly 1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3}' because of the following reason: debugRuntimeClasspath uses version 1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3

    > Could not resolve io.flutter:flutter_embedding_debug:{strictly 1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3}.
        Required by:
            project :flutter_downloader
        > Cannot find a version of 'io.flutter:flutter_embedding_debug' that satisfies the version constraints:
            Dependency path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:1.0.0-ad3d868e0d174f5a8447b011bd714d18a5f2abec'
            Dependency path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3'
            Constraint path 'vn.hunghd.flutterdownloader:flutter_downloader:1.0-SNAPSHOT' --> 'io.flutter:flutter_embedding_debug:{strictly 1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3}' because of the following reason: debugRuntimeClasspath uses version 1.0.0-a4ff2c53d84ca78702bc3be6c7ef0788ffca01d3


    * Try:
    > Run with --stacktrace option to get the stack trace.
    > Run with --info or --debug option to get more log output.
    > Run with --scan to get full insights.

    * Get more help at https://help.gradle.org

    BUILD FAILED in 21s
    Exception: Gradle task assembleDebug failed with exit code 1
    Exited
</details>

File: android/build.gradle
Comment (rekire) [reply android/build.gradle:22]: This was a test to provide a dependency so that I don't need to edit the code in two Android Studio instances, it makes trouble as I see so I'll revert this

File: android/build.gradle
Comment (rekire) [reply android/build.gradle:54]: This was a test to provide a dependency so that I don't need to edit the code in two Android Studio instances, it makes trouble as I see so I'll revert this

File: example/android/app/build.gradle
Comment (rekire) [reply example/android/app/build.gradle:36]: fine

File: example/android/app/build.gradle
Comment (rekire) [reply example/android/app/build.gradle:48]: I tough it would be more common to add that line, but I personally like it more too to remove it. We might should think about a common format to avoid this.

File: android/src/main/AndroidManifest.xml
Comment (rekire) [reply android/src/main/AndroidManifest.xml:10]: With the downgrade to 32 this is not required so sure...

File: android/build.gradle
Comment (rekire) [reply android/build.gradle:27]: Yes this is required. This compiles the kotlin code to the JVM bytecode.

**Comment by rekire** [comment]
@bartekpacia could you please check it again?

**Comment by bartekpacia** [comment]
I tested it a bit and works fine. I'd really like to have some integration tests though.

File: example/android/app/build.gradle
Comment (bartekpacia) [inline example/android/app/build.gradle:48]: nit: remove this line

File: example/android/app/build.gradle
Comment (bartekpacia) [inline example/android/app/build.gradle:56]: nit: re-add newline

File: example/android/build.gradle
Comment (bartekpacia) [inline example/android/build.gradle:8]: Let's bump this in a separate PR as well :)

File: example/android/build.gradle
Comment (bartekpacia) [inline example/android/build.gradle:29]: nit: re-add newline

**Review by bartekpacia — COMMENTED** [review]
Just 4 small things so we keep the diff to the minimum.

Regarding code style and linting, I'd like to introduce `ktlint` and `detekt` after we merge this PR.

**Comment by rekire** [comment]
Well I tried to keep the formatting consistent, so you want to to have inconsistent just to keep the diff minimal?

**Comment by bartekpacia** [comment]
yeah, I'll reformat everything with ktlint afterwards :)

**Comment by rekire** [comment]
For such small white space changes you should have the permission to change that yourself. There is no need, out of my view, to wait for my change :-)

However I think that none of those tools would format the gradle files automatically.

**Review by bartekpacia — APPROVED** [review]

**Comment by bartekpacia** [comment]
Thanks a lot for your contribution! :)

**Pull request merged**
